### PR TITLE
Add end-to-end tests for the CLI

### PR DESCRIPTION
**Pull request by kmaziarz**
So far our search CLI was not being tested in CI at all, while the single-step evaluation CLI was only covered with some unit tests using synthetic models/data. This made it quite easy to break the CLI, or break the model wrappers in a way that is evident only when e.g. using them end-to-end during search.

This PR adds full end-to-end tests for all of our CLI entry points, using all single-step models (apart from GLN), and - in the case of the search CLI - all main search algorithm classes. This increases coverage of `cli/{main, search}.py` from 0% to ~90%, and of `cli/eval_single_step.py` from 75% to 85%. Additionally, testing end-to-end search means added coverage of e.g. utilities for dumping routes into `*.pdf` files, which were also not being tested. Overall, this PR improves test coverage from 81% to 91%.

These new tests uncovered two issues related to returning probability values: one in Chemformer (introduced in recent refactoring, evident when running search), and one in RootAligned (introduced a long time ago, evident when running single-step evaluation with saving of results turned on); both of these are now also fixed as part of this PR. Finally, I also randomly uncovered a quirk of MEGAN, which causes it to fail if one sets `device="cpu"` despite GPU being available (this is because there are a lot of `torch.cuda.is_available()` calls in MEGAN code that "contradict" the fact that we've moved the model to CPU). For now, I just raise an explicit error in that case, as it's likely not a very useful scenario anyway.

### CODE REVIEW

**Review by AustinT — CHANGES_REQUESTED** [review]
Overall looks very good Kris. Thanks for writing this test! I have two high-level comments which I think you should change though.

1. I don't like the idea of linking to somebody's dropbox (also left a comment about this). I think it is important to change this (so I marked this PR as "request changes")
2. I notice you put all the tests in `test_models.py`. Perhaps it would make more sense to put them in `test_cli.py` instead? fixtures which are shared between the two files could be put in `conftest.py`, as is already done in some of the search tests.

**Comment by kmaziarz** [comment]
> I notice you put all the tests in `test_models.py`. Perhaps it would make more sense to put them in `test_cli.py` instead?

Would you put this new file in the same directory i.e. `tests/reaction_prediction/inference`? Alternatively these could be placed in a new directory `tests/cli` but then I assume the fixtures would have to live really high up in this directory hierarchy to be usable by both tests?

**Comment by AustinT** [comment]
> Would you put this new file in the same directory i.e. `tests/reaction_prediction/inference`? Alternatively these could be placed in a new directory `tests/cli` but then I assume the fixtures would have to live really high up in this directory hierarchy to be usable by both tests?

Good question: I guess it CLI tests could be top level? But I agree this makes the fixtures harder to find. Feel free to leave it as it is.

File: syntheseus/tests/reaction_prediction/inference/test_models.py [hunk 1]
```diff
@@ -53,6 +68,50 @@ def model(request) -> ExternalBackwardReactionModel:
...
+def data_dir() -> Generator[Path, None, None]:
+    with tempfile.TemporaryDirectory() as raw_tempdir:
+        tempdir = Path(raw_tempdir)
+
+        # Download the raw USPTO-50K data released by the authors of GLN.
+        uspto50k_zip_path = tempdir / "uspto50k.zip"
+        urllib.request.urlretrieve(
+            "https://www.dropbox.com/sh/6ideflxcakrak10/AADN-TNZnuGjvwZYiLk7zvwra/schneider50k?dl=1",
+            uspto50k_zip_path,
+        )
```
[the remark this answers is not in the document]
Comment (kmaziarz) [reply syntheseus/tests/reaction_prediction/inference/test_models.py]: Makes sense; I replaced this with our own mirror.

**Comment by kmaziarz** [comment]
> I guess CLI tests could be top level? But I agree this makes the fixtures harder to find.

I realized the CLI tests do not share fixtures with the models tests; they only share the logic for skipping the test if the single-step models are not installed. I now extracted that logic to a separate utility, which allowed for splitting out the CLI tests into `tests/cli/test_cli.py`.

**Review by AustinT — APPROVED** [review]
Looks good to me Kris!

**Pull request merged**